File: app/src/main/java/vn/loitp/app/activity/customviews/layout/coordinatorlayout/CoordinatorLayoutWithImageViewActivity.java
package vn.loitp.app.activity.customviews.layout.coordinatorlayout;

import android.graphics.Matrix;
import android.os.Bundle;
import android.widget.ImageView;

import com.annotation.LayoutId;
import com.core.base.BaseFontActivity;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;

import vn.loitp.app.R;

@LayoutId(R.layout.activity_coordinator_layout)
public class CoordinatorLayoutWithImageViewActivity extends BaseFontActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final ImageView imageView = findViewById(R.id.img_hero);
        AppBarLayout appBarLayout = findViewById(R.id.app_bar);
        appBarLayout.addOnOffsetChangedListener((appBarLayout1, verticalOffset) -> {
            Matrix matrix = new Matrix(imageView.getImageMatrix());

            //get image's width and height
            final int dwidth = imageView.getDrawable().getIntrinsicWidth();
            final int dheight = imageView.getDrawable().getIntrinsicHeight();

            //get view's width and height
            final int vwidth = imageView.getWidth() - imageView.getPaddingLeft() - imageView.getPaddingRight();
            int vheight = imageView.getHeight() - imageView.getPaddingTop() - imageView.getPaddingBottom();

            float scale;
            float dx = 0, dy = 0;
            float parallaxMultiplier = ((CollapsingToolbarLayout.LayoutParams) imageView.getLayoutParams()).getParallaxMultiplier();

            //maintain the image's aspect ratio depending on offset
            if (dwidth * vheight > vwidth * dheight) {
                vheight += (verticalOffset); //calculate view height depending on offset
                scale = (float) vheight / (float) dheight; //calculate scale
                dx = (vwidth - dwidth * scale) * 0.5f; //calculate x value of the center point of scaled drawable
                dy = -verticalOffset * (1 - parallaxMultiplier); //calculate y value by compensating parallaxMultiplier
            } else {
                scale = (float) vwidth / (float) dwidth;
                dy = (vheight - dheight * scale) * 0.5f;
            }

            int currentWidth = Math.round(scale * dwidth); //calculate current intrinsic width of the drawable

            if (vwidth <= currentWidth) { //compare view width and drawable width to decide, should we scale more or not
                matrix.setScale(scale, scale);
                matrix.postTranslate(Math.round(dx), Math.round(dy));
                imageView.setImageMatrix(matrix);
            }
        });
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return getClass().getSimpleName();
    }

}

File: app/src/main/java/vn/loitp/app/activity/customviews/layout/coordinatorlayout/CoordinatorLayoutWithImageViewActivity.kt
package vn.loitp.app.activity.customviews.layout.coordinatorlayout

import android.graphics.Matrix
import android.os.Bundle
import com.annotation.IsFullScreen
import com.annotation.LayoutId
import com.annotation.LogTag
import com.core.base.BaseFontActivity
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.AppBarLayout.OnOffsetChangedListener
import com.google.android.material.appbar.CollapsingToolbarLayout
import kotlinx.android.synthetic.main.activity_coordinator_layout.*
import vn.loitp.app.R
import kotlin.math.roundToInt

@LayoutId(R.layout.activity_coordinator_layout)
@LogTag("CoordinatorLayoutWithImageViewActivity")
@IsFullScreen(false)
class CoordinatorLayoutWithImageViewActivity : BaseFontActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        appBar.addOnOffsetChangedListener(OnOffsetChangedListener { _: AppBarLayout?, verticalOffset: Int ->
            val matrix = Matrix(imgHero.imageMatrix)

            //get image's width and height
            val dWidth = imgHero.drawable.intrinsicWidth
            val dHeight = imgHero.drawable.intrinsicHeight

            //get view's width and height
            val vWidth = imgHero.width - imgHero.paddingLeft - imgHero.paddingRight
            var vHeight = imgHero.height - imgHero.paddingTop - imgHero.paddingBottom
            val scale: Float
            var dx = 0f
            val dy: Float
            val parallaxMultiplier = (imgHero.layoutParams as CollapsingToolbarLayout.LayoutParams).parallaxMultiplier

            //maintain the image's aspect ratio depending on offset
            if (dWidth * vHeight > vWidth * dHeight) {
                vHeight += verticalOffset //calculate view height depending on offset
                scale = vHeight.toFloat() / dHeight.toFloat() //calculate scale
                dx = (vWidth - dWidth * scale) * 0.5f //calculate x value of the center point of scaled drawable
                dy = -verticalOffset * (1 - parallaxMultiplier) //calculate y value by compensating parallaxMultiplier
            } else {
                scale = vWidth.toFloat() / dWidth.toFloat()
                dy = (vHeight - dHeight * scale) * 0.5f
            }
            val currentWidth = (scale * dWidth).roundToInt() //calculate current intrinsic width of the drawable
            if (vWidth <= currentWidth) { //compare view width and drawable width to decide, should we scale more or not
                matrix.setScale(scale, scale)
                matrix.postTranslate(dx.roundToInt().toFloat(), dy.roundToInt().toFloat())
                imgHero.imageMatrix = matrix
            }
        })
    }
}
